source/server.go

```diff
@@ -0,0 +1,101 @@
+// Copyright © 2023 Meroxa, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package source
+
+import (
+	"context"
+	"fmt"
+	"sync"
+
+	"github.com/conduitio-labs/conduit-connector-grpc-server/fromproto"
+	pb "github.com/conduitio-labs/conduit-connector-grpc-server/proto/v1"
+	sdk "github.com/conduitio/conduit-connector-sdk"
+	"gopkg.in/tomb.v2"
+)
+
+type Server struct {
+	pb.UnimplementedSourceServiceServer
+
+	RecordCh    chan sdk.Record
+	Teardown    chan bool
+	streamMutex sync.Mutex
+	openContext context.Context
+	tomb        *tomb.Tomb
+	stream      pb.SourceService_StreamServer
+}
+
+func NewServer(ctx context.Context) *Server {
+	return &Server{
+		RecordCh: make(chan sdk.Record),
+		// buffering channel, so Teardown won't be blocked until the signal is received
+		Teardown:    make(chan bool, 1),
+		openContext: ctx,
+		tomb:        &tomb.Tomb{},
+	}
+}
+
+func (s *Server) Stream(stream pb.SourceService_StreamServer) error {
+	s.streamMutex.Lock()
+	if s.stream != nil {
+		s.streamMutex.Unlock()
+		return fmt.Errorf("only one client connection is supported")
+	}
+	s.stream = stream
+	s.streamMutex.Unlock()
+
+	// spawn a go routine to receive records from client
+	s.tomb.Go(s.recvRecords)
```

Comment:
I asked myself the same question but figured it's still simpler with a tomb, cuz I'm gonna need a channel to get errors from the go routine and a way to stop the go routine in case the context is canceled, and that's easily done by killing and listening to a tomb, also the tomb is used to track the lifecycle of one or more goroutines, so why not 👀
ofc I'm not against it either, and it works both ways, so lemme know what you think

Reply:
We need it after all, because receiving messages from a proto stream doesn't take a context.